File: C#_module/SuperGLU/SuperGLU/xAPILearnLogger.cs
using Newtonsoft.Json.Linq;
using System;
using System.IO;
using System.Collections.Generic;
using TinCan;


namespace SuperGLU
{
    public class xAPILearnLogger: BaseLearnLogger
    {
        string mbox = "mailto:SMART-E@ict.usc.edu";
        string activityURI = "http://id.tincanapi.com/activity/";

        /* ---------------- constructor, getter and setter --------------- */
        public xAPILearnLogger(string userId = "", string userName = ""): base()
        {
            this.userId = userId;
            this.userName = userName;
            this.url = new Uri("https://github.com/GeneralizedLearningUtilities/SuperGLU/");
        }

        public void SetUserId(string userId)
        {
            this.userId = userId;
        }

        public void SetUserName(string userName)
        {
            this.userName = userName;
        }

        /* ---------------------- create xAPI object ---------------------- */
        // create actor
        private Agent CreateActor(string name, string openid, string mbox)
        {
            Agent actor = new Agent();
            actor.name = name;
            actor.openid = openid;
            actor.mbox = mbox;

            return actor;
        }

        // create agent account, used as an instructor agent in context
        private AgentAccount CreateAgentAccount(string name, string homePage)
        {
            AgentAccount agentAccount = new AgentAccount();
            agentAccount.name = name;
            agentAccount.homePage = new Uri(homePage);

            return agentAccount;
        }

        // create verb
        private Verb CreateCompletedVerb()
        {
            Verb verb = new Verb();
            verb.id = new Uri("http://activitystrea.ms/schema/1.0/complete");
            verb.display = new LanguageMap();
            verb.display.Add("en-US", "completed");

            return verb;
        }

        private Verb CreateStartedVerb()
        {
            Verb verb = new Verb();
            verb.id = new Uri("http://activitystrea.ms/schema/1.0/start");
            verb.display = new LanguageMap();
            verb.display.Add("en-US", "started");

            return verb;
        }

        private Verb CreateTerminiatedVerb()
        {
            Verb verb = new Verb();
            verb.id = new Uri("http://activitystrea.ms/schema/1.0/terminate");
            verb.display = new LanguageMap();
            verb.display.Add("en-US", "terminated");

            return verb;
        }

        private Verb CreatePresentedVerb()
        {
            Verb verb = new Verb();
            verb.id = new Uri("http://activitystrea.ms/schema/1.0/present");
            verb.display = new LanguageMap();
            verb.display.Add("en-US", "presented");

            return verb;
        }

        // create activity
        private Activity CreateSession(string activityId, string name, string description)
        {
            Activity activity = new Activity();
            activity.id = this.activityURI + activityId;
            activity.definition = new ActivityDefinition();
            activity.definition.name = new LanguageMap(new Dictionary<string, string>(){ { "en-US", name } });
            activity.definition.description = new LanguageMap(new Dictionary<string, string>() { { "en-US", description } });
            activity.definition.type = new Uri("http://id.tincanapi.com/activitytype/tutor-session");

            return activity;
        }

        private Activity CreateTopic(string activityId, string name, string description)
        {

            Activity activity = new Activity();
            activity.id = this.activityURI + activityId;
            activity.definition = new ActivityDefinition();
            activity.definition.name = new LanguageMap(new Dictionary<string, string>() { { "en-US", name } });
            activity.definition.description = new LanguageMap(new Dictionary<string, string>() { { "en-US", description } });
            activity.definition.type = new Uri("http://id.tincanapi.com/activitytype/topic");

            return activity;
        }

        private Activity CreateLesson(string activityId, string name, string description)
        {

            Activity activity = new Activity();
            activity.id = this.activityURI + activityId;
            activity.definition = new ActivityDefinition();
            activity.definition.name = new LanguageMap(new Dictionary<string, string>() { { "en-US", name } });
            activity.definition.description = new LanguageMap(new Dictionary<string, string>() { { "en-US", description } });
            activity.definition.type = new Uri("http://id.tincanapi.com/activitytype/lesson");

            return activity;
        }

        private Activity CreateSublesson(string activityId, string name, string description)
        {

            Activity activity = new Activity();
            activity.id = this.activityURI + activityId;
            activity.definition = new ActivityDefinition();
            activity.definition.name = new LanguageMap(new Dictionary<string, string>() { { "en-US", name } });
            activity.definition.description = new LanguageMap(new Dictionary<string, string>() { { "en-US", description } });
            activity.definition.type = new Uri("http://id.tincanapi.com/activitytype/sublesson");

            return activity;
        }

        private Activity CreateTask(string activityId, string name, string description)
        {

            Activity activity = new Activity();
            activity.id = this.activityURI + activityId;
            activity.definition = new ActivityDefinition();
            activity.definition.name = new LanguageMap(new Dictionary<string, string>() { { "en-US", name } });
            activity.definition.description = new LanguageMap(new Dictionary<string, string>() { { "en-US", description } });
            activity.definition.type = new Uri("http://id.tincanapi.com/activitytype/task");

            return activity;
        }

        private Activity CreateStep(string activityId, string name, string description)
        {

            Activity activity = new Activity();
            activity.id = this.activityURI + activityId;
            activity.definition = new ActivityDefinition();
            activity.definition.name = new LanguageMap(new Dictionary<string, string>() { { "en-US", name } });
            activity.definition.description = new LanguageMap(new Dictionary<string, string>() { { "en-US", description } });
            activity.definition.type = new Uri("http://id.tincanapi.com/activitytype/step");

            return activity;
        }

        private Activity CreateVideo(string activityId= "http://activitystrea.ms/schema/1.0/video", string name= "video", string description= "Video content of any kind")
        {

            Activity activity = new Activity();
            activity.id = this.activityURI + activityId;
            activity.definition = new ActivityDefinition();
            activity.definition.name = new LanguageMap(new Dictionary<string, string>() { { "en-US", name } });
            activity.definition.description = new LanguageMap(new Dictionary<string, string>() { { "en-US", description } });
            activity.definition.type = new Uri("http://id.tincanapi.com/activitytype/video");

            return activity;
        }

        // TODO: activity tree class need to be implemented
        private Activity CreateActivityTree(string activityId = "http://activitystrea.ms/schema/1.0/activityTree", string name = "activityTree", string description = "Activity tree (not implemented yet)")
        {
            Activity activity = new Activity();
            activity.id = this.activityURI + activityId;
            activity.definition = new ActivityDefinition();
            activity.definition.name = new LanguageMap(new Dictionary<string, string>() { { "en-US", name } });
            activity.definition.description = new LanguageMap(new Dictionary<string, string>() { { "en-US", description } });
            activity.definition.type = new Uri("http://id.tincanapi.com/activitytype/activityTree");

            return activity;
        }

        // create result
        private Result CreateResult(string response = "", bool? success = null)
        {
            Result result = new Result();
            if (response != "")
            {
                result.response = response;
            }
            if (success != null)
            {
                result.success = success;
            }

            return result;
        }

        // create context
        private Context CreateContext(JObject contextJson)
        {   
            // instructor agent
            AgentAccount agentAccount = CreateAgentAccount("instructorAgent", "http://instructorHomepage.com");

            Context context = new Context();
            context.registration = Guid.NewGuid();   // globally unique indentifier
            context.instructor = new Agent(new JObject { { "account", agentAccount.ToJObject()} });
            context.extensions = new TinCan.Extensions(contextJson);    // 'Extensions' is ambiguous reference between 'Newtonsoft.Json' and 'TinCan'
            context.contextActivities = new ContextActivities();

            return context;
        }

        /* ------------------- construct xAPI statement ------------------- */
        // start
        public void SendStartSession(string activityId, string name, string description, JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateStartedVerb();
            Activity activity = CreateSession(activityId, name, description);
            Result result = CreateResult(response: "User started a new Session");
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        public void SendStartLesson(string activityId, string name, string description, JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateStartedVerb();
            Activity activity = CreateLesson(activityId, name, description);
            Result result = CreateResult(success: true);
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        public void SendStartSublesson(string activityId, string name, string description, JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateStartedVerb();
            Activity activity = CreateSublesson(activityId, name, description);
            Result result = CreateResult(success: true);
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        public void SendStartTask(string activityId, string name, string description, JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateStartedVerb();
            Activity activity = CreateTask(activityId, name, description);
            Result result = CreateResult(success: true);
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        public void SendStartStep(string activityId, string name, string description, JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateStartedVerb();
            Activity activity = CreateStep(activityId, name, description);
            Result result = CreateResult(success: true);
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        // terminiate or complete
        public void SendTerminiatedSession(JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateTerminiatedVerb();
            Activity activity = CreateActivityTree();   // activity tree class need to be implemented, this is a dummy method
            Result result = CreateResult();
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        public void SendCompletedLesson(JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateCompletedVerb();
            Activity activity = CreateActivityTree();   // activity tree class need to be implemented, this is a dummy method
            Result result = CreateResult();
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        public void SendCompletedSublesson(JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateCompletedVerb();
            Activity activity = CreateActivityTree();   // activity tree class need to be implemented, this is a dummy method
            Result result = CreateResult();
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        public void SendCompletedTask(JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateCompletedVerb();
            Activity activity = CreateActivityTree();   // activity tree class need to be implemented, this is a dummy method
            Result result = CreateResult();
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        /* work in progress. currently requires custom_score_URI and custom_score but not all applications will have this.
           If a raw_score is provided then a max_score must be provided too.
           Might want to provide more detailed information relating to the knowledge components involved in the step.
        */
        public void SendCompletedStep(string choice, string customScoreURI, double customScore, JObject contextJson, double rawScore = -1, double maxScore = -1, double minScore = 0, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreateCompletedVerb();
            Activity activity = CreateActivityTree();
            Result result = new Result();
            Context context = CreateContext(contextJson);
            TinCan.Extensions extensions = new TinCan.Extensions(new JObject { {customScoreURI, customScore} });

            // handle different result
            if (rawScore != -1)
            {
                result.response = choice;
                result.score = new Score(new JObject { { "raw", rawScore }, { "min", minScore }, {"max", maxScore} });
                result.extensions = extensions;
            } else
            {
                result.response = choice;
                result.extensions = extensions;
            }

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }


        public void SendPresentedVideo(JObject contextJson, DateTime? timestamp = null)
        {
            Agent actor = CreateActor(this.userName, this.userId, this.mbox);
            Verb verb = CreatePresentedVerb();
            Activity activity = CreateVideo(); 
            Result result = CreateResult();
            Context context = CreateContext(contextJson);

            if (timestamp == null)
            {
                timestamp = GetTimeStamp();
            }

            JObject statementJson = new JObject
            {
                { "actor", actor.ToJObject() },
                { "verb", verb.ToJObject() },
                { "object", activity.ToJObject() },
                { "result", result.ToJObject() },
                { "context", context.ToJObject() },
                { "timestamp", timestamp }

            };

            Statement statement = new Statement(statementJson);
            this.SendLoggingMessage(statement);
        }

        private void SendLoggingMessage(Statement statement)
        {
            string path = Directory.GetCurrentDirectory();
            // write to log file
            using (System.IO.StreamWriter logFile =
                new System.IO.StreamWriter(path + "\\xAPILog.json", true))
                {
                    logFile.WriteLine(statement.ToJSON());
                }
        }
    }

   

}

File: java_maven/src/main/java/edu/usc/ict/superglu/services/logging/XAPILearnLogger.java
package edu.usc.ict.superglu.services.logging;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;

import edu.usc.ict.superglu.core.BaseService;
import edu.usc.ict.superglu.core.Message;
import edu.usc.ict.superglu.core.MessagingGateway;
import edu.usc.ict.superglu.core.MessagingVerbConstants;
import edu.usc.ict.superglu.core.SpeechActEnum;
import gov.adlnet.xapi.model.Account;
import gov.adlnet.xapi.model.Activity;
import gov.adlnet.xapi.model.ActivityDefinition;
import gov.adlnet.xapi.model.Agent;
import gov.adlnet.xapi.model.Context;
import gov.adlnet.xapi.model.ContextActivities;
import gov.adlnet.xapi.model.Result;
import gov.adlnet.xapi.model.Score;
import gov.adlnet.xapi.model.Statement;
import gov.adlnet.xapi.model.Verb;

public class XAPILearnLogger extends BaseService {

	public static String BASE_URI ="https://github.com/GeneralizedLearningUtilities/SuperGLU/";
	public static String SESSION_TYPE = "http://id.tincanapi.com/activitytype/tutor-session";
	public static String LESSON_TYPE = "http://adlnet.gov/expapi/activities/lesson";
	public static String SUBLESSON_TYPE = BASE_URI + "sublesson";
	public static String TASK_TYPE = "http://activitystrea.ms/schema/1.0/task";
	public static String STEP_TYPE = "http://id.tincanapi.com/activitytype/step";

	//# VERBS
	public static String START_URI = "http://activitystrea.ms/schema/1.0/start";
	public static String COMPLETED_URI = "http://activitystrea.ms/schema/1.0/complete";
	public static String TERMINATED_URI = "http://activitystrea.ms/schema/1.0/terminate";
	public static String WATCHED_URI = "http://activitystrea.ms/schema/1.0/watch";

	//# ACTIVITY DEFINITION EXTENSION
	public static String ACTIVITY_ID_URI = BASE_URI + "activityid";

	//# CONTEXT EXTENSIONS
	public static String RECOVERED_URI = BASE_URI + "recovered";
	public static String MISSING_COMPS_URI = BASE_URI + "missingCompletions";
	public static String ACTIVITY_TREE_URI = BASE_URI + "serialized_activitytree";

	private MessagingGateway gateway;
	private String userId;
	private String userName;
	private String homePage;
	private String mboxHost;
	private String errorLogName ="xap_learn_logger_errorLog.txt";
	private float secondsAfterLastTimeStamp = 1;
	private ActivityTree activityTree;
	
	public static SimpleDateFormat timestampFormat = new SimpleDateFormat("yyyy-MM-ddTHH:mm:ss.SSSZ");
	
	
	public XAPILearnLogger(MessagingGateway gateway, String userId, String userName, String homePage, String mboxHost)
	{
		super();
		this.gateway = gateway;
		this.userId = userId;
		this.userName = userName;
		this.homePage = homePage;
		this.mboxHost = mboxHost;
		this.activityTree = new ActivityTree();
	}
	
	
	public void setUserId(String userId)
	{
		this.userId = userId;
	}
	
	
	public void setUserName(String userName)
	{
		this.userName = userName;
	}
	
	
	public Date getTimestamp()
	{
		return new Date();
	}
	
	
	//VERBS
	
	public Verb createCompletedVerb()
	{
		HashMap<String, String> display = new HashMap<>();
		display.put("en-US", "completed");
		Verb result = new Verb(COMPLETED_URI, display);
		return result;
	}
	
	
	public Verb createStartedVerb()
	{
		HashMap<String, String> display = new HashMap<>();
		display.put("en-US", "started");
		Verb result = new Verb(START_URI, display);
		return result;
	}
	
	
	public Verb createTerminatedVerb()
	{
		HashMap<String, String> display = new HashMap<>();
		display.put("en-US", "terminated");
		Verb result = new Verb(TERMINATED_URI, display);
		return result;
	}
	
	
	public Verb createWatchedVerb()
	{
		HashMap<String, String> display = new HashMap<>();
		display.put("en-US", "watched");
		Verb result = new Verb(WATCHED_URI, display);
		return result;
	}
	
	
	// ACTIVITIES
	
	public Activity createSession(String activityId, String name, String description)
	{
		HashMap<String, String> nameMap = new HashMap<>();
		nameMap.put("en-US", name);
		HashMap<String, String> descriptionMap = new HashMap<>();
		descriptionMap.put("en-US", description);
		ActivityDefinition definition = new ActivityDefinition(nameMap, descriptionMap);
		definition.setType(SESSION_TYPE);
		HashMap<String, JsonElement> extensions = new HashMap<>();
		extensions.put(ACTIVITY_ID_URI, new JsonPrimitive(UUID.randomUUID().toString()));
		definition.setExtensions(extensions);
		Activity result = new Activity(activityId, definition);
		return result;
		
	}
	
	
	public Activity createLesson(String activityId, String name, String description)
	{
		HashMap<String, String> nameMap = new HashMap<>();
		nameMap.put("en-US", name);
		HashMap<String, String> descriptionMap = new HashMap<>();
		descriptionMap.put("en-US", description);
		ActivityDefinition definition = new ActivityDefinition(nameMap, descriptionMap);
		definition.setType(LESSON_TYPE);
		HashMap<String, JsonElement> extensions = new HashMap<>();
		extensions.put(ACTIVITY_ID_URI, new JsonPrimitive(UUID.randomUUID().toString()));
		definition.setExtensions(extensions);
		Activity result = new Activity(activityId, definition);
		return result;
	}
	
	
	public Activity createSubLesson(String activityId, String name, String description)
	{
		HashMap<String, String> nameMap = new HashMap<>();
		nameMap.put("en-US", name);
		HashMap<String, String> descriptionMap = new HashMap<>();
		descriptionMap.put("en-US", description);
		ActivityDefinition definition = new ActivityDefinition(nameMap, descriptionMap);
		definition.setType(SUBLESSON_TYPE);
		HashMap<String, JsonElement> extensions = new HashMap<>();
		extensions.put(ACTIVITY_ID_URI, new JsonPrimitive(UUID.randomUUID().toString()));
		definition.setExtensions(extensions);
		Activity result = new Activity(activityId, definition);
		return result;
	}
	
	
	public Activity createTask(String activityId, String name, String description)
	{
		HashMap<String, String> nameMap = new HashMap<>();
		nameMap.put("en-US", name);
		HashMap<String, String> descriptionMap = new HashMap<>();
		descriptionMap.put("en-US", description);
		ActivityDefinition definition = new ActivityDefinition(nameMap, descriptionMap);
		definition.setType(TASK_TYPE);
		HashMap<String, JsonElement> extensions = new HashMap<>();
		extensions.put(ACTIVITY_ID_URI, new JsonPrimitive(UUID.randomUUID().toString()));
		definition.setExtensions(extensions);
		Activity result = new Activity(activityId, definition);
		return result;
	}
	
	
	public Activity createStep(String activityId, String name, String description)
	{
		HashMap<String, String> nameMap = new HashMap<>();
		nameMap.put("en-US", name);
		HashMap<String, String> descriptionMap = new HashMap<>();
		descriptionMap.put("en-US", description);
		ActivityDefinition definition = new ActivityDefinition(nameMap, descriptionMap);
		definition.setType(STEP_TYPE);
		HashMap<String, JsonElement> extensions = new HashMap<>();
		extensions.put(ACTIVITY_ID_URI, new JsonPrimitive(UUID.randomUUID().toString()));
		definition.setExtensions(extensions);
		Activity result = new Activity(activityId, definition);
		return result;
	}
	
	
	public Activity createVideo()
	{
		HashMap<String, String> nameMap = new HashMap<>();
		nameMap.put("en-US", "video");
		HashMap<String, String> descriptionMap = new HashMap<>();
		descriptionMap.put("en-US", "Video content of any kind");
		ActivityDefinition definition = new ActivityDefinition(nameMap, descriptionMap);
		HashMap<String, JsonElement> extensions = new HashMap<>();
		extensions.put(ACTIVITY_ID_URI, new JsonPrimitive(UUID.randomUUID().toString()));
		definition.setExtensions(extensions);
		Activity result = new Activity("http://activitystrea.ms/schema/1.0/video", definition);
		return result;
	}
	
	
	public Agent createAgent()
	{
		Agent result;
		
		if(this.mboxHost != null)
		{
			result = new Agent(this.userName, "mailto:" + this.userName + "@" + this.mboxHost);
		}
		else if(this.homePage != null)
		{
			Account account = new Account(this.userId, this.homePage);
			result = new Agent(this.userName, account);
		}
		else
		{
			result = new Agent(this.userName, "");
		}
		
		return result;
	}
	
	
	//Starting and Stopping
	
	public void sendStartSession(String activityID, String name, String description, HashMap<String, JsonElement> contextDict, Date timestamp)
	{
		Activity activity = this.createSession(activityID, name, description);
		Agent actor = this.createAgent();
		
		this.activityTree.enterActivity(null, activity);
		Context context = this.addContext(contextDict);
		
		if(timestamp == null)
			timestamp = this.getTimestamp();
		
		Statement statement = new Statement(actor, this.createStartedVerb(), activity);
		statement.setResult(null);
		String timestampAsString = timestampFormat.format(timestamp);
		statement.setTimestamp(timestampAsString);
		statement.setContext(context);
		
		this.sendLoggingMessage(statement);
		
	}
	
	
	public void sendStartLesson(String activityID, String name, String description, HashMap<String, JsonElement> contextDict, Date timestamp)
	{
		Activity activity = this.createLesson(activityID, name, description);
		Agent actor = this.createAgent();
		
		this.activityTree.enterActivity(null, activity);
		Context context = this.addContext(contextDict);
		
		if(timestamp == null)
			timestamp = this.getTimestamp();
		
		Statement statement = new Statement(actor, this.createStartedVerb(), activity);
		statement.setResult(null);
		String timestampAsString = timestampFormat.format(timestamp);
		statement.setTimestamp(timestampAsString);
		statement.setContext(context);
		
		this.sendLoggingMessage(statement);		
	}
	
	
	public void sendStartSubLesson(String activityID, String name, String description, HashMap<String, JsonElement> contextDict, Date timestamp)
	{
		Activity activity = this.createSubLesson(activityID, name, description);
		Agent actor = this.createAgent();
		
		this.activityTree.enterActivity(null, activity);
		Context context = this.addContext(contextDict);
		
		if(timestamp == null)
			timestamp = this.getTimestamp();
		
		Statement statement = new Statement(actor, this.createStartedVerb(), activity);
		statement.setResult(null);
		String timestampAsString = timestampFormat.format(timestamp);
		statement.setTimestamp(timestampAsString);
		statement.setContext(context);
		
		this.sendLoggingMessage(statement);
		
	}
	
	
	public void sendStartTask(String activityID, String name, String description, HashMap<String, JsonElement> contextDict, Date timestamp)
	{
		Activity activity = this.createTask(activityID, name, description);
		Agent actor = this.createAgent();
		
		this.activityTree.enterActivity(null, activity);
		Context context = this.addContext(contextDict);
		
		if(timestamp == null)
			timestamp = this.getTimestamp();
		
		Statement statement = new Statement(actor, this.createStartedVerb(), activity);
		statement.setResult(null);
		String timestampAsString = timestampFormat.format(timestamp);
		statement.setTimestamp(timestampAsString);
		statement.setContext(context);
		
		this.sendLoggingMessage(statement);
		
	}
	
	
	public void sendStartStep(String activityID, String name, String description, HashMap<String, JsonElement> contextDict, Date timestamp)
	{
		Activity activity = this.createStep(activityID, name, description);
		Agent actor = this.createAgent();
		
		this.activityTree.enterActivity(null, activity);
		Context context = this.addContext(contextDict);
		
		if(timestamp == null)
			timestamp = this.getTimestamp();
		
		Statement statement = new Statement(actor, this.createStartedVerb(), activity);
		statement.setResult(null);
		String timestampAsString = timestampFormat.format(timestamp);
		statement.setTimestamp(timestampAsString);
		statement.setContext(context);
		
		this.sendLoggingMessage(statement);
		
	}
	
	
	public HashMap<String, JsonElement> createFakeContextDict()
	{
		HashMap<String, JsonElement> result = new HashMap<>();
		
		result.put(RECOVERED_URI, new JsonPrimitive("activity"));
		return result;
	}
	
	
	public void sendTerminatedSession(HashMap<String, JsonElement> contextDict, Date timestamp)
	{
		Agent actor = this.createAgent();
		
		Activity activity = this.activityTree.findCurrentActivity();
		
		boolean missingData = false;
		
		if(timestamp == null)
		{
			timestamp = this.getTimestamp();
		}
	}
	
	
	public void sendCompletedLesson(HashMap<String, JsonElement> contextDict, Date timestamp, boolean fake)
	{
		Agent actor = this.createAgent();
		Activity activity = this.activityTree.findCurrentActivity();
		Context context = this.addContext(contextDict);
		
		this.activityTree.exitActivity();
		
		if(timestamp == null)
		{
			timestamp = this.getTimestamp();
		}
		Verb verb;
		if(fake)
		{
			verb = this.createTerminatedVerb();
		}
		else
		{
			verb = this.createCompletedVerb();
		}
		
		Statement statement = new Statement(actor, verb, activity);
		statement.setResult(null);
		statement.setContext(context);
		statement.setTimestamp( timestampFormat.format(timestamp));
		
		this.sendLoggingMessage(statement);
	}
	
	
	public void sendCompletedSublesson(HashMap<String, JsonElement> contextDict, Date timestamp, boolean fake)
	{
		Agent actor = this.createAgent();
		Activity activity = this.activityTree.findCurrentActivity();
		Context context = this.addContext(contextDict);
		
		this.activityTree.exitActivity();
		
		if(timestamp == null)
		{
			timestamp = this.getTimestamp();
		}
		Verb verb;
		if(fake)
		{
			verb = this.createTerminatedVerb();
		}
		else
		{
			verb = this.createCompletedVerb();
		}
		
		Statement statement = new Statement(actor, verb, activity);
		statement.setResult(null);
		statement.setContext(context);
		statement.setTimestamp( timestampFormat.format(timestamp));
		
		this.sendLoggingMessage(statement);
	}
	
	
	public void sendCompletedTask(HashMap<String, JsonElement> contextDict, Date timestamp, boolean fake)
	{
		Agent actor = this.createAgent();
		Activity activity = this.activityTree.findCurrentActivity();
		Context context = this.addContext(contextDict);
		
		this.activityTree.exitActivity();
		
		if(timestamp == null)
		{
			timestamp = this.getTimestamp();
		}
		Verb verb;
		if(fake)
		{
			verb = this.createTerminatedVerb();
		}
		else
		{
			verb = this.createCompletedVerb();
		}
		
		Statement statement = new Statement(actor, verb, activity);
		statement.setResult(null);
		statement.setContext(context);
		statement.setTimestamp( timestampFormat.format(timestamp));
		
		this.sendLoggingMessage(statement);
	}
	
	
	public void sendCompletedStep(String choice, HashMap<String, JsonElement> contextDict, JsonObject resultExtDict, float rawScore, float maxScore, float minScore, Date timestamp, boolean fake)
	{
		Agent actor = this.createAgent();
		
		Score score;
		if(rawScore != -1)
		{		

			score = new Score();
			score.setRaw(rawScore);
			score.setMin(minScore);
			score.setMax(maxScore);
		}
		else
		{
			score = null;
		}
		
		Result result = new Result();
		result.setResponse(choice);
		result.setScore(score);
		result.setExtensions(resultExtDict);
		
		Activity activity = this.activityTree.findCurrentActivity();
		Context context = this.addContext(contextDict);
		
		this.activityTree.exitActivity();
		
		if(timestamp == null)
		{
			timestamp = this.getTimestamp();
		}
		Verb verb;
		if(fake)
		{
			verb = this.createTerminatedVerb();
		}
		else
		{
			verb = this.createCompletedVerb();
		}
		
		Statement statement = new Statement(actor, verb, activity);
		statement.setResult(result);
		statement.setContext(context);
		statement.setTimestamp( timestampFormat.format(timestamp));
		
		this.sendLoggingMessage(statement);		
	}
	
	
	public void sendWatchedVideo(HashMap<String, JsonElement> contextDict, Date timestamp)
	{
		Agent actor = createAgent();
		
		Activity activity = createVideo();
		
		this.activityTree.enterActivity(null, activity);
		Context context = this.addContext(contextDict);
		this.activityTree.exitActivity();
		
		if(timestamp == null)
		{
			timestamp = this.getTimestamp();
		}
		
		Statement statement = new Statement(actor, this.createWatchedVerb(), activity);
		statement.setResult(null);
		statement.setContext(context);
		statement.setTimestamp(timestampFormat.format(timestamp));
		this.sendLoggingMessage(statement);
	}
	
	
	
	
	
	//Utility functions
	
	public Context addContext(HashMap<String, JsonElement> contextDict)
	{
		contextDict.put(ACTIVITY_TREE_URI, new JsonPrimitive(this.activityTree.activityTreeToSimple()));
		
		List<Activity> myGrouping = this.activityTree.convertPathToGrouping();
		Activity myParent = this.activityTree.findParentActivity();
		
		Context result;
		
		if(myGrouping.size() == 0 && myParent== null)
		{
			result = new Context();
			result.setExtensions(contextDict);
		}
		else
		{
			ContextActivities myContextActivities = new ContextActivities();
			
			myContextActivities = new ContextActivities();
			ArrayList<Activity> parentList = new ArrayList<>();
			parentList.add(myParent);
			myContextActivities.setParent(parentList);
			if(myGrouping.size() > 0)
			{
				myContextActivities.setGrouping((ArrayList<Activity>) myGrouping);
			}
			
			result = new Context();
			result.setExtensions(contextDict);
			result.setContextActivities(myContextActivities);
		}
		
		return result;
	}
	
	
	public void addContext()
	{
		HashMap<String, JsonElement> contextDict = new HashMap<>();
		this.addContext(contextDict);
	}
	
	public void sendLoggingMessage(Statement statement)
	{
		Message message = new Message();
		message.setActor("logger");
		message.setVerb(MessagingVerbConstants.XAPI_LOG_VERB);
		message.setObj(null);
		message.setResult(statement.serialize().toString());
		message.setSpeechAct(SpeechActEnum.INFORM_ACT);
		this.sendMessage(message);
	}
			
}
